### ISSUE TITLE
sm.tl.spatial_lda stores result in adata.uns not adata.obs

### ISSUE THREAD

**Issue by marinabroz** [labsyspharm/scimap#24]
Hello,

I am trying to run sm.tl.spatial_lda with my adata object but for some reason the spatial_lda results are stored in adata.uns and not adata.obs. This is causing some errors with the clustering downstream using sm.tl.cluster. Any help would be appreciated!

I am working in Python v3.9.

After running the code: 

adata= sm.tl.spatial_lda(adata, x_coordinate='X', y_coordinate='Y', phenotype='celltype', method='radius', radius=30, knn=10, imageid='UniqueID', num_motifs=10, random_state=0, subset=None, label='spatial_lda')

adata 

AnnData object with n_obs × n_vars = 79308 × 34
    obs: 'Unnamed: 0', 'X', 'Y', 'Area', 'celltype', 'TLSType', 'UniqueID'
    uns: 'spatial_lda', 'spatial_lda_probability', 'spatial_lda_model'

![Scimap](https://user-images.githubusercontent.com/88685061/166714448-425ea26c-1180-4504-924b-3848fc72826d.png)

**Comment by ajitjohnson** [comment]
Hi @marinabroz thank you for bringing up this issue. I see that there is an issue with the documentation and a documentation page is missing. 

The results of running the `lda` analysis are supposed to be saved in `adata.uns`, however subsequently you would cluster it with
`sm.tl.spatial_cluster` function and not the regular `sm.tl.cluster` function. I see that this can be confusing and will try to converge into one function later on. 

Please try the following and the documentation is up now at : https://scimap.xyz/All%20Functions/B.%20Tools/sm.tl.spatial_cluster/

```
adata = sm.tl.spatial_cluster (adata, df_name='spatial_lda', method = 'kmeans', k=15, label='spatial_lda_kmeans')
```

**Comment by marinabroz** [comment]
Thank you for the quick reply! Using sm.tl.spatial_cluster worked for the clustering. Thanks again!

**Comment by ajitjohnson** [comment]
umm I guess that depends on the machine. It should not take that long I guess maybe a few minutes but not 30 mins. can you try including the `subsample=1000` parameter to see if the size is the issue? The `cluster_plots` is a wrapper and I do not generally use it as much. If you are familiar with `scanpy` you could directly use their umap and heatmap functions. The data is directly compatible and would not need any alterations. 

check out the umap plots I generate here: https://scimap.xyz/tutorials/2-scimap-tutorial-cell-phenotyping/

**Comment by marinabroz** [comment]
Thanks, I think I can see why its not working.  It seems that sm.tl.spatial_cluster returns NaN values for all of the cells. Have you encountered this before? When I look into the spatial_lda object there are many zero values for the weights, is this acceptable or could it be causing this issue?

![Scimap2](https://user-images.githubusercontent.com/88685061/166809816-5397d92a-1e5f-4f06-acbe-4ddb0694b3eb.png)

**Comment by ajitjohnson** [comment]
hmm, that is difficult to say as those are latent variables and not immediately interpretable. What kind of data are you using? Have you tried varying the `radius` parameter? Depending on the pixel size, you might have to change this. An easier first step would be to use the `knn` method than the `radius` method. 

In any case it does not explain why you would get `Nan` in your clustering result. Can you export `adata.uns['spatial_lda']` and share it with me?

**Comment by marinabroz** [comment]
I am working with imaging mass cytometry data, and my tissues are tumor and immune cells. I have attached the data from uns[spatial_lda] here. For this run I have increased the radius to 100. Thank you!


[spatial_lda.csv](https://github.com/labsyspharm/scimap/files/8627964/spatial_lda.csv)

**Comment by ajitjohnson** [comment]
Thank you. Just running your raw `spatial_lda.csv` through the functions works fine. Unfortunately, I would need your entire `adata object` to debug any further. If you are sending it over please save the `h5ad` file with everything that you ran right before the step where you run the `spatial_cluster` function.

**Comment by marinabroz** [comment]
I appreciate your help! This is my first time working in python so I'm struggling a bit. I have attached the h5 file. I did recieve this error when attempting to `adata.write`  , let me know if this is acceptable. Thank you!

NotImplementedError: Failed to write value for uns/spatial_lda_model, since a writer for type <class 'gensim.models.ldamulticore.LdaMulticore'> has not been implemented yet

[adata.zip](https://github.com/labsyspharm/scimap/files/8628220/adata.zip)
.

**Comment by ajitjohnson** [comment]
No problem. Hmm it works fine for me. Can you `pip install scimap --upgrade` to make sure you are on the latest version and try again? 

Also, i would suggest grouping your cell-types into larger categories so that you have fewer cell-types to run through these algorithms. You have too many variables with limited field of view/ number of cells and so it will be more interpretable or powerful if you can group your cell types into larger categories. Once you have identified the spatial domains, you can always look into it with finer cell type classification.

**Comment by marinabroz** [comment]
Hi Ajit, sorry for the late reply! Another member in my lab played around with it, and she was getting NaN at first as well. However, when she ran the following we were able to return cluster numbers:

adata=sm.tl.spatial_cluster(adata, random_state=0, df_name='spatial_lda')

Also I appreciate your input on the groupings of cells, we gave that a try as well and I'm pretty happy with the results. Thank you for all of your help!

**Comment by ajitjohnson** [comment]
No problem. Glad it worked out :)

**Comment by daniannr** [comment]
Hi Ajit, I'm opening this issue again because I'm have the same errors as above. I ran `sm.tl.spatial_count` followed by `sm.tl.spatial_cluster` and resulted in all NaN in the cluster assignments. My adata has 16 tissue samples and 200k cells with around 15k cells in each sample. 

I'm not sure if it's an issue with k for because I tried using k=5-100. Here's what I ran:
`adata = sm.tl.spatial_count(adata, phenotype='finalcelltype',imageid="samplename", method='knn',knn=50, label='spatial_count')`
`adata=sm.tl.spatial_cluster(adata,  method='kmeans', k=6, label='neigh_kmeans', df_name='spatial_count')
`
Hope you can help. Thanks!

**Comment by ajitjohnson** [comment]
can you subsample your data? maybe a total of 10K cells? attach it here with the exact lines for me to reproduce the error? thanks!

**Comment by daniannr** [comment]
Thanks for your response! File is above. It has 9600 cells. This is the code I ran for neighborhood analysis using `spatial_count`:
`adata_subsampled = sm.tl.spatial_count(adata_subsampled, phenotype='Celesta_finalcelltype',imageid="samplename", method='knn',knn=10, label='spatial_count')`
`adata_subsampled=sm.tl.spatial_cluster(adata_subsampled,  method='kmeans', k=6, label='neigh_kmeans', df_name='spatial_count')
`

I tried running the two lines again and it worked for the subset data where clusters were not NaN. However, when I tried using `sm.tl.spatial_lda`, it's the same issue. From watching the progress, it looks like it tries to run my entire script for some reason? Not sure what is happening.

Also, there are 16 images in the adata. The image IDs are under the "samplename" metadata column. Is `spatial_cluster` ran per imageid or across all samples/images?

**Comment by daniannr** [comment]
[adatasubset_logZ_scimap.h5ad.zip](https://github.com/user-attachments/files/19372203/adatasubset_logZ_scimap.h5ad.zip)

Thanks for your response! File is above. It has 9600 cells. This is the code I ran for neighborhood analysis using `spatial_count`:
`adata_subsampled = sm.tl.spatial_count(adata_subsampled, phenotype='Celesta_finalcelltype',imageid="samplename", method='knn',knn=10, label='spatial_count')`
`adata_subsampled=sm.tl.spatial_cluster(adata_subsampled,  method='kmeans', k=6, label='neigh_kmeans', df_name='spatial_count')
`

I tried running the two lines again and it worked for the subset data where clusters were not NaN. However, when I tried using `sm.tl.spatial_lda`, it's the same issue. From watching the progress, it looks like it tries to run my entire script for some reason? Not sure what is happening.

Also, there are 16 images in the adata. The image IDs are under the "samplename" metadata column. Is `spatial_cluster` ran per imageid or across all samples/images?
